mlflow/pyspark/ml/__init__.py

```diff
@@ -107,24 +120,129 @@ def _get_estimator_info_tags(estimator):
+
+    instance_name = gen_instance_name(instance)
+    if isinstance(instance, Pipeline):
+        sub_stages = []
+        for sub_stage in instance.getStages():
+            sub_stages.append(_gen_instance_metadata_recursively(
+                sub_stage, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+            ))
+        return {'name': instance_name, 'stages': sub_stages}
+    elif isinstance(instance, OneVsRest):
+        classifier = _gen_instance_metadata_recursively(
+            instance.getClassifier(),
+            uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+        )
+        return {'name': instance_name, 'classifier': classifier}
+    elif _is_parameter_search_estimator(instance):
+        param_search_estimators.append(instance)
+        evaluator = instance.getEvaluator()
+        tuned_estimator = instance.getEstimator()
+        return {
+            'name': instance_name,
+            'evaluator': _gen_instance_metadata_recursively(
+                evaluator, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+            ),
+            'tuned_estimator': _gen_instance_metadata_recursively(
+                tuned_estimator, uid2name_map, cls2count_map, cls2first_uid_map,
+                param_search_estimators
+            )
+        }
+    else:
+        return instance_name
 
 
-def _get_instance_param_map_recursively(instance, level):
+def _gen_estimator_metadata(estimator):
+    """
+    Returns an AutologgingEstimatorMetadata object.
+    The AutologgingEstimatorMetadata object includes:
+     - hierarchy: the hierarchy of the estimator, it will expand
+         pipeline/meta estimator/param tuning estimator
+     - uid2name_map: a map of `uid` -> `name`, each nested instance uid will be
+         mapped to a fixed name. The naming rule is using `{class_name}` if the
+         instance type occurs once, or `{class_name}_{index}` if the instance type occurs
+         multiple times. The index order is in line with depth-first traversing.
+     - param_search_estimators: a list includes all param search estimators in the
+         hierarchy tree.
+    """
+    uid2name_map = {}
+    cls2count_map = {}
+    cls2first_uid_map = {}
+    param_search_estimators = []
+
+    hierarchy = _gen_instance_metadata_recursively(
+        estimator, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+    )
+    for cls, first_uid in cls2first_uid_map.items():
+        if cls2count_map[cls] > 1:
+            # change the first stage name from "{class_name}" to "{class_name}_1"
+            uid2name_map[first_uid] = f'{cls}_1'
+
+    for i in range(len(param_search_estimators)):
+        name = param_search_estimators[i]
+        if name in cls2count_map and cls2count_map[name] > 1:
+            # change the estimator name from "{class_name}" to "{class_name}_1"
+            param_search_estimators[i] = f'{name}_1'
```

Comment:
I'm a bit confused about why we use `f'{name}_1'` instead of `f'{name}_i'`. Am I missing something or is there an error in the code? If it's a bug, can we add a test case? If it isn't a bug, can we add an line comment explaining why this is something we want to do?

Reply:
Refactor related logic to make it easier to read.